vscode/src/createProject.ts

```diff
@@ -149,4 +165,191 @@ export async function initProjectCreator(context: vscode.ExtensionContext) {
+    // Ensure we have something
+    if (!from || !to) throw "Invalid arguments";
+
+    // Trim trailing slashes (even from the root "/" case)
+    if (from.endsWith("/")) from = from.slice(0, -1);
+    if (to.endsWith("/")) to = to.slice(0, -1);
+
+    // Break both paths into their components
+    const fromParts = from.split("/");
+    const toParts = to.split("/");
+
+    // Remove the common beginning of the paths
+    while (fromParts[0] === toParts[0]) {
+      fromParts.shift();
+      toParts.shift();
+    }
+
+    // Add a .. for each remaining part in the from path
+    let result = "";
+    while (fromParts.length) {
+      result += "../";
+      fromParts.shift();
+    }
+    // Add the remaining path from the to path
+    result += toParts.join("/");
+    if (result.endsWith("/")) {
+      result = result.slice(0, -1);
+    }
+    return result;
+  }
+
+  context.subscriptions.push(
+    vscode.commands.registerCommand(
+      "qsharp-vscode.addProjectReference",
+      async (qsharpJsonUri: vscode.Uri | undefined) => {
+        // If called from the content menu qsharpJsonUri will be the full qsharp.json uri
+        // If called from the command palette is will be undefined, so use the active editor
+        log.info("addProjectReference called with", qsharpJsonUri);
+
+        qsharpJsonUri =
+          qsharpJsonUri ?? vscode.window.activeTextEditor?.document.uri;
+        if (!qsharpJsonUri) {
+          log.error(
+            "addProjectReference called, but argument or active editor is not qsharp.json",
+          );
+          return;
+        }
+
+        log.debug("Adding project reference to ", qsharpJsonUri.path);
+
+        // First, verify the qsharp.json can be opened and is a valid json file
+        const qsharpJsonDoc =
+          await vscode.workspace.openTextDocument(qsharpJsonUri);
+        if (!qsharpJsonDoc) {
+          log.error("Unable to open the qsharp.json file at ", qsharpJsonDoc);
+          return;
+        }
+        const qsharpJsonDir = vscode.Uri.joinPath(qsharpJsonUri, "..");
+
+        let manifestObj: any = {};
+        try {
+          manifestObj = JSON.parse(qsharpJsonDoc.getText());
+        } catch (err: any) {
+          await vscode.window.showErrorMessage(
+            `Unable to parse the contents of ${qsharpJsonUri.path}`,
+          );
+          return;
+        }
+
+        // Find all the other Q# projects in the workspace
+        const projectFiles = (
+          await vscode.workspace.findFiles("**/qsharp.json")
+        ).filter((file) => file.toString() !== qsharpJsonUri.toString());
+
+        const projectChoices: Array<{ name: string; ref: Dependency }> = [];
+
+        projectFiles.forEach((file) => {
+          const dirName = file.path.slice(0, -"/qsharp.json".length);
+          const relPath = getRelativeDirPath(qsharpJsonDir.path, dirName);
+          projectChoices.push({
+            name: dirName.slice(dirName.lastIndexOf("/") + 1),
```

Comment:
Isn't there a chance this could add libraries with directory names with special characters or whitespace and then try to use those as aliases? I haven't actually tried that, and I'm not sure what we do to sanitize the name chosen in the json. It's possible this could cause unintended behavior, so we should test what the failure handling in the dependency path is for spaces.

Reply:
Yeah, I was thinking we might need to do some 'normalizing'. Even without whitespace or "special characters", it's quite common to have a dash in a directory name (it added `df-chemistry` for example, which is invalid).

We could expect the user to fix up in truly oddball cases, but at least handle the common case of dashes and spaces (maybe just replace with a `_`).